src/he/pmtud.c

```diff
@@ -0,0 +1,287 @@
+      return he_internal_pmtud_base_confirmed(conn);
+    case HE_PMTUD_STATE_SEARCHING:
+      // Current probe acked
+      if(conn->pmtud_probing_size >= MAX_PLPMTU) {
+        // Search completed if the MAX_PLPMTU acked
+        return he_internal_pmtud_search_completed(conn);
+      } else {
+        // Increment probing size and send next probe
+        uint16_t probe_size = conn->pmtud_probing_size;
+        probe_size +=
+            (conn->pmtud_is_using_big_step) ? PMTUD_PROBE_BIG_STEP : PMTUD_PROBE_SMALL_STEP;
+        conn->pmtud_probing_size = MIN(probe_size, MAX_PLPMTU);
+        he_internal_pmtud_send_probe(conn, conn->pmtud_probing_size);
+      }
+      break;
+    case HE_PMTUD_STATE_ERROR:
+      // Base PMTU probe acked, enter Searching state now
+      return he_internal_pmtud_base_confirmed(conn);
+    default:
+      // Do nothing
+      break;
+  }
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_handle_probe_timeout(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+
+  // Try again with the same probe size
+  if(conn->pmtud_probe_count < MAX_PROBES) {
+    return he_internal_pmtud_send_probe(conn, conn->pmtud_probing_size);
+  }
+
+  // PROBE_COUNT reaches MAX_PROBES, decide what to do based on state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+      // Unable to confirm the base PMTU, entering error state.
+      return he_internal_pmtud_confirm_base_failed(conn);
+    case HE_PMTUD_STATE_SEARCHING:
+      if(conn->pmtud_is_using_big_step) {
+        // Try probing with small step
+        conn->pmtud_is_using_big_step = false;
+        conn->pmtud_probing_size -= PMTUD_PROBE_BIG_STEP;
+        conn->pmtud_probing_size += PMTUD_PROBE_SMALL_STEP;
+        return he_internal_pmtud_send_probe(conn, conn->pmtud_probing_size);
+      } else {
+        // Search completed
+        // Set the probing size to the previous successful one
+        assert(conn->pmtud_probing_size > conn->pmtud_base);
+        conn->pmtud_probing_size -= PMTUD_PROBE_SMALL_STEP;
+        return he_internal_pmtud_search_completed(conn);
+      }
+      break;
+    case HE_PMTUD_STATE_SEARCH_COMPLETE:
+      // Black hole detected
+      return he_internal_pmtud_blackhole_detected(conn);
+    default:
+      // Do nothing in DISABLED / ERROR states
+      break;
+  }
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_start_base_probing(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_DISABLED:
+    case HE_PMTUD_STATE_SEARCHING:
+    case HE_PMTUD_STATE_SEARCH_COMPLETE:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_BASE);
+
+  // Initialize PMTUD internal state
+  conn->pmtud_base = MAX(MIN_PLPMTU, conn->effective_pmtu);
```

Comment:
I am not sure whether we need to consider `conn->effective_pmtu` here.

Thinking about this case:
- User initiates PMTU discovery and it succeeds at 1350
- User changes internet connection which reduces the MTU
- User triggers another discovery, and now we start with 1350 as base and it will fail. We will move the state machine to he_internal_pmtud_confirm_base_failed() and no way to recover.

We need to either start from same `MIN_PLPMTU` value or update `he_internal_pmtud_handle_probe_timeout` to handle this case. I would prefer to use `MIN_PLPMTU` cause it is simple to visualize.

Please let me know if I miss something

Reply:
Yes, you're right. There's a follow up ticket to handle the scenarios you described, so for now I'll fix the code to check MIN_PLPMTU only.